[WIP] LOR solvers: device and p-refinement [lor-solvers-p-ref]

pazner opened the pull request:
Extends the `LORDiscretization` and `LORSolver` functionality:

* Works with *p*-refined finite element spaces (including *hp*-refinement).
* Compatibility with hypre's GPU solvers

The main difference is that the DOF permutation (required for Nédélec and Raviart-Thomas spaces since the LOR and HO DOFs have different numberings) is built into to the `P` and `R` matrices of the finite element space (even for conforming meshes), so the `LORSolver` class does not need to perform any permutations.

Preliminary testing looks good using hypre's GPU solvers with #1492.

pazner commented:
@jakubcerveny, what is the status of p-refinement for ND and RT elements? I am trying to figure out what I should be expecting to work in my tests, and what is not yet supported.

jakubcerveny commented:
I haven't tried ND and RT yet. In particular the collections are not extended for variable order yet (that should be fairly easy though). In principle however, the hp interpolation only depends on FE::GetTransferMatrix, so it should eventually "just" work (fingers crossed).

pazner commented:
Thanks for letting me know. So what exactly happens when I call `SetElementOrder` on an ND or RT finite element space? It doesn't crash, but is something internally going wrong (since the collections are not extended for variable order)?

jakubcerveny commented:
It will not crash right away, you can set element orders in the space as you like, but when you try to update the space (it gets constructed again with the new orders in Construct()), the first call to FECollection::GetNumDof (or the like) will try to Clone() the collection for the new order and that should produce a message if the collection is not ready for variable orders.

pazner commented:
I am able to call `FiniteElementSpace::Update` and it still doesn't give an error message. For concreteness, I inserted the following snippet in ex3 and ex4:
```c++
   mesh->EnsureNCMesh();
   for (int i = 0; i < mesh->GetNE(); i++)
   {
      int el_order = 1 + (rand()%3);
      fespace->SetElementOrder(i, el_order);
   }
   fespace->Update(false);
```

jakubcerveny commented:
Hm, you are right, I thought the ND and RT collection don't have Clone() yet, but it turns out you added the methods in a57c6ca3 :-) I guess it's not debugged yet, I'd start with simple two-quad meshes and check that the DOFs on multi-order edges make sense...

pazner commented:
This should be ready for review. The main improvement is that the LOR solvers should now be completely compatible with hypre's AMG, AMS, and ADS solvers on the GPU. I added a `--device` flag to the parallel version of the LOR solvers miniapp (in serial it uses either Gauss-Seidel or a direct solver, so I don't think device runs are relevant in that context).

Support for _p_- and _hp_-refined spaces is also added. This changes slightly how the `P` matrices are used. Instead of performing DOF permutations in `Mult`, now the DOF permutation is built into the `P` matrix, which should also improve performance.

crdohrm commented:
Looks like you've made a lot of good progress on the GPU front. Thank you for keeping me in the loop.

brendankeith reviewed (approved):
@pazner helped me to test the PR. LGTM

tzanio reviewed (approved):
Apologies for the delayed review @pazner, this looks great!

tzanio commented:
Merged in `next` for testing...

pazner commented:
Just pushed a small bug fix in c7856f70eade40c24d99b4f243c8bb397d6d2106, this should be re-merged in next when convenient

v-dobrev commented:
Re-merged in `next` for testing.
